In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns
from sklearn.model_selection import train_test_split, GridSearchCV 
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import cross_val_score 
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier, AdaBoostClassifier
from sklearn.metrics import accuracy_score

In [2]:
df = pd.read_csv('winemag-data-130k-v2 2.csv')
df.head()

Unnamed: 0   country                                        description  \
0           0     Italy  Aromas include tropical fruit, broom, brimston...   
1           1  Portugal  This is ripe and fruity, a wine that is smooth...   
2           2        US  Tart and snappy, the flavors of lime flesh and...   
3           3        US  Pineapple rind, lemon pith and orange blossom ...   
4           4        US  Much like the regular bottling from 2012, this...   

                          designation  points  price           province  \
0                        Vulkà Bianco      87    NaN  Sicily & Sardinia   
1                            Avidagos      87   15.0              Douro   
2                                 NaN      87   14.0             Oregon   
3                Reserve Late Harvest      87   13.0           Michigan   
4  Vintner's Reserve Wild Child Block      87   65.0             Oregon   

              region_1           region_2         taster_name  \
0                 Etna                NaN       Kerin O’Keefe   
1                  NaN                NaN          Roger Voss   
2    Willamette Valley  Willamette Valley        Paul Gregutt   
3  Lake Michigan Shore                NaN  Alexander Peartree   
4    Willamette Valley  Willamette Valley        Paul Gregutt   

  taster_twitter_handle                                              title  \
0          @kerinokeefe                  Nicosia 2013 Vulkà Bianco  (Etna)   
1            @vossroger      Quinta dos Avidagos 2011 Avidagos Red (Douro)   
2           @paulgwine       Rainstorm 2013 Pinot Gris (Willamette Valley)   
3                   NaN  St. Julian 2013 Reserve Late Harvest Riesling ...   
4           @paulgwine   Sweet Cheeks 2012 Vintner's Reserve Wild Child...   

          variety               winery  
0     White Blend              Nicosia  
1  Portuguese Red  Quinta dos Avidagos  
2      Pinot Gris            Rainstorm  
3        Riesling           St. Julian  
4      Pinot Noir         Sweet Cheeks

In [3]:
len(df)

129971

In [4]:
df.variety.describe()

count         129970
unique           707
top       Pinot Noir
freq           13272
Name: variety, dtype: object

In [5]:
nullvals = df.variety.isnull().values
len(nullvals)

129971

## Find out how many null values exist and get rid of the features with the most null vals

In [6]:

df.isna().sum()

Unnamed: 0                   0
country                     63
description                  0
designation              37465
points                       0
price                     8996
province                    63
region_1                 21247
region_2                 79460
taster_name              26244
taster_twitter_handle    31213
title                        0
variety                      1
winery                       0
dtype: int64

In [7]:
# region_2 and twitter handle have the most null values so dropping the column
df_cleaned1 = df.drop(['Unnamed: 0','region_2','taster_twitter_handle'],axis=1)
df_cleaned1

country                                        description  \
0             Italy  Aromas include tropical fruit, broom, brimston...   
1          Portugal  This is ripe and fruity, a wine that is smooth...   
2                US  Tart and snappy, the flavors of lime flesh and...   
3                US  Pineapple rind, lemon pith and orange blossom ...   
4                US  Much like the regular bottling from 2012, this...   
5             Spain  Blackberry and raspberry aromas show a typical...   
6             Italy  Here's a bright, informal red that opens with ...   
7            France  This dry and restrained wine offers spice in p...   
8           Germany  Savory dried thyme notes accent sunnier flavor...   
9            France  This has great depth of flavor with its fresh ...   
10               US  Soft, supple plum envelopes an oaky structure ...   
11           France  This is a dry wine, very spicy, with a tight, ...   
12               US  Slightly reduced, this wine offers a chalky, t...   
13            Italy  This is dominated by oak and oak-driven aromas...   
14               US  Building on 150 years and six generations of w...   
15          Germany  Zesty orange peels and apple notes abound in t...   
16        Argentina  Baked plum, molasses, balsamic vinegar and che...   
17        Argentina  Raw black-cherry aromas are direct and simple ...   
18            Spain  Desiccated blackberry, leather, charred wood a...   
19               US  Red fruit aromas pervade on the nose, with cig...   
20               US  Ripe aromas of dark berries mingle with ample ...   
21               US  A sleek mix of tart berry, stem and herb, alon...   
22            Italy  Delicate aromas recall white flower and citrus...   
23               US  This wine from the Geneseo district offers aro...   
24            Italy  Aromas of prune, blackcurrant, toast and oak c...   
25               US  Oak and earth intermingle around robust aromas...   
26            Italy  Pretty aromas of yellow flower and stone fruit...   
27            Italy  Aromas recall ripe dark berry, toast and a whi...   
28            Italy  Aromas suggest mature berry, scorched earth, a...   
29               US  Clarksburg is becoming a haven for Chenin Blan...   
...             ...                                                ...   
129941           US  A Chardonnay with an unusual companion, 8% Sém...   
129942           US  This is classic in herbaceous aromas and flavo...   
129943        Italy  A blend of Nero d'Avola and Syrah, this convey...   
129944       Israel  Deep garnet in the glass, this has a nose of b...   
129945           US  Hailing from one of the more popular vineyards...   
129946      Germany  Plump, clingy peach and honey notes are cut wi...   
129947        Italy  A blend of 65% Cabernet Sauvignon, 30% Merlot ...   
129948    Argentina  Raspberry and cassis aromas are fresh and upri...   
129949           US  There's no bones about the use of oak in this ...   
129950           US  This opens with herbaceous dollops of thyme an...   
129951       France  Hugely spicy this rich wine is described as sw...   
129952           US  This Zinfandel from the eastern section of Nap...   
129953  New Zealand  Roughly two-thirds Cabernet and one-third Merl...   
129954  New Zealand  One of the more characterful Pinot Gris for th...   
129955  New Zealand  Like Dog Point's 2011 Chardonnay, this wine is...   
129956  New Zealand  The blend is 44% Merlot, 33% Cabernet Sauvigno...   
129957        Spain  Lightly baked berry aromas vie for attention w...   
129958  New Zealand  This blend of Cabernet Sauvignon-Merlot and Ca...   
129959       France  The granite soil of the Brand Grand Cru vineya...   
129960     Portugal  Fresh and fruity, this is full of red cherry f...   
129961        Italy  Intense aromas of wild cherry, baking spice, t...   
129962        Italy  Blackberry, cassis, grilled herb and toasted a...   
129963       Isr

In [8]:
df_cleaned2 = df_cleaned1.dropna()

In [9]:
len(df_cleaned2) 

54170

In [10]:
len(df) - len(df_cleaned2) #number of lost data points - lost over half our data points

75801

In [11]:
df_cleaned2

country                                        description  \
3              US  Pineapple rind, lemon pith and orange blossom ...   
4              US  Much like the regular bottling from 2012, this...   
5           Spain  Blackberry and raspberry aromas show a typical...   
6           Italy  Here's a bright, informal red that opens with ...   
9          France  This has great depth of flavor with its fresh ...   
10             US  Soft, supple plum envelopes an oaky structure ...   
16      Argentina  Baked plum, molasses, balsamic vinegar and che...   
17      Argentina  Raw black-cherry aromas are direct and simple ...   
18          Spain  Desiccated blackberry, leather, charred wood a...   
20             US  Ripe aromas of dark berries mingle with ample ...   
22          Italy  Delicate aromas recall white flower and citrus...   
23             US  This wine from the Geneseo district offers aro...   
24          Italy  Aromas of prune, blackcurrant, toast and oak c...   
25             US  Oak and earth intermingle around robust aromas...   
26          Italy  Pretty aromas of yellow flower and stone fruit...   
28          Italy  Aromas suggest mature berry, scorched earth, a...   
35             US  As with many of the Erath 2010 vineyard design...   
42         France  This is a festive wine, with soft, ripe fruit ...   
49         France  Soft and fruity, this is a generous, ripe wine...   
53         France  Fruity and lightly herbaceous, this has fine t...   
60             US  Syrupy and dense, this wine is jammy in plum a...   
61          Italy  This densely hued wine has aromas of black plu...   
62             US  The aromas are brooding, with notes of barrel ...   
63         France  This fat, yeasty Champagne is comprised predom...   
64             US  There are intriguing touches to the nose of th...   
67             US  A blend of Merlot and Cabernet Franc, this win...   
69         France  This cooperative, based in Aÿ, has benefited f...   
71             US  Big oak defines this robustly dense and extrac...   
72          Italy  Aromas of black-skinned fruit, leather, underb...   
73             US  Juicy plum, raspberry and pencil lead lead the...   
...           ...                                                ...   
129902      Spain  Dry apple and lemon aromas are clean and one s...   
129903     France  A fine value for a northern Rhône of this qual...   
129904         US  A blend of Sauvignon Blanc, Sauvignon Gris and...   
129909         US  Hailing from an extreme coastal vineyard, this...   
129912         US  From the Ranch House block of the famous viney...   
129913         US  This is taut and dense, and requires time and ...   
129915         US  The third vintage of Mith is the best yet, com...   
129917  Argentina  This dark, meaty Malbec has some Agrelo fruit ...   
129919         US  This ripe, rich, almost decadently thick wine ...   
129921  Argentina  There is a select group of under-$20 Malbecs f...   
129925  Argentina  A lively, well-made blend of Tempranillo, Malb...   
129926         US  This pure Syrah from Reininger's estate vineya...   
129928      Spain  This represents Torres' highest effort in Prio...   
129930      Spain  Arguably on the heavy and rich side, but if yo...   
129932  Argentina  Andeluna's top wines tend to be ripe and plump...   
129936     France  A blend from several premier cru vineyards in ...   
129938  Argentina  Compared to the regular 2006 Malbec from Chaka...   
129943      Italy  A blend of Nero d'Avola and Syrah, this convey...   
129945         US  Hailing from one of the more popular vineyards...   
129947      Italy  A blend of 65% Cabernet Sauvignon, 30% Merlot ...   
129948  Argentina  Raspberry and cassis aromas are fresh and upri...   
129949         US  There's no bones about the use of oak in this ...   
129950         US  This opens with herbaceous dollops of thyme an...   
129951     France  Hugely spicy this rich wine is 

In [12]:
list(df_cleaned2.columns.values)

['country',
 'description',
 'designation',
 'points',
 'price',
 'province',
 'region_1',
 'taster_name',
 'title',
 'variety',
 'winery']

In [13]:
#double check no null vals
df_cleaned2.isna().sum()

country        0
description    0
designation    0
points         0
price          0
province       0
region_1       0
taster_name    0
title          0
variety        0
winery         0
dtype: int64

In [14]:
rwo_df = pd.read_excel('Types of Wine - O, R, W.xlsx')
rwo_df.head()

variety   color
0  Alvar Roxo  Orange
1       Ar110  Orange
2        Ar99  Orange
3  Barbarossa  Orange
4   Barbaroux  Orange

In [15]:
df_cleaned2.describe(include= 'all')

country                                        description designation  \
count    54170                                              54170       54170   
unique       7                                              49537       23943   
top         US  Stalky aromas suggest hay and green herbs, wit...     Reserve   
freq     25888                                                  3        1159   
mean       NaN                                                NaN         NaN   
std        NaN                                                NaN         NaN   
min        NaN                                                NaN         NaN   
25%        NaN                                                NaN         NaN   
50%        NaN                                                NaN         NaN   
75%        NaN                                                NaN         NaN   
max        NaN                                                NaN         NaN   

              points         price    province              region_1  \
count   54170.000000  54170.000000       54170                 54170   
unique           NaN           NaN          63                  1004   
top              NaN           NaN  California  Columbia Valley (WA)   
freq             NaN           NaN       13998                  2488   
mean       88.966550     39.209341         NaN                   NaN   
std         2.955188     36.154076         NaN                   NaN   
min        80.000000      4.000000         NaN                   NaN   
25%        87.000000     20.000000         NaN                   NaN   
50%        89.000000     30.000000         NaN                   NaN   
75%        91.000000     48.000000         NaN                   NaN   
max       100.000000   2013.000000         NaN                   NaN   

       taster_name                                        title     variety  \
count        54170                                        54170       54170   
unique          17                                        48994         434   
top     Roger Voss  Segura Viudas NV Extra Dry Sparkling (Cava)  Pinot Noir   
freq          9237                                            8        6736   
mean           NaN                                          NaN         NaN   
std            NaN                                          NaN         NaN   
min            NaN                                          NaN         NaN   
25%            NaN                                          NaN         NaN   
50%            NaN                                          NaN         NaN   
75%            NaN                                          NaN         NaN   
max            NaN                                          NaN         NaN   

                winery  
count            54170  
unique            8583  
top     Columbia Crest  
freq               157  
mean               NaN  
std                NaN  
min                NaN  
25%                NaN  
50%                NaN  
75%                NaN  
max                NaN

In [16]:
df_cleaned3 = df_cleaned2.merge(rwo_df, on='variety', how='left')
df_cleaned3

country                                        description  \
0             US  Pineapple rind, lemon pith and orange blossom ...   
1             US  Much like the regular bottling from 2012, this...   
2          Spain  Blackberry and raspberry aromas show a typical...   
3          Italy  Here's a bright, informal red that opens with ...   
4         France  This has great depth of flavor with its fresh ...   
5             US  Soft, supple plum envelopes an oaky structure ...   
6      Argentina  Baked plum, molasses, balsamic vinegar and che...   
7      Argentina  Raw black-cherry aromas are direct and simple ...   
8          Spain  Desiccated blackberry, leather, charred wood a...   
9             US  Ripe aromas of dark berries mingle with ample ...   
10         Italy  Delicate aromas recall white flower and citrus...   
11            US  This wine from the Geneseo district offers aro...   
12         Italy  Aromas of prune, blackcurrant, toast and oak c...   
13            US  Oak and earth intermingle around robust aromas...   
14         Italy  Pretty aromas of yellow flower and stone fruit...   
15         Italy  Aromas suggest mature berry, scorched earth, a...   
16            US  As with many of the Erath 2010 vineyard design...   
17        France  This is a festive wine, with soft, ripe fruit ...   
18        France  Soft and fruity, this is a generous, ripe wine...   
19        France  Fruity and lightly herbaceous, this has fine t...   
20            US  Syrupy and dense, this wine is jammy in plum a...   
21         Italy  This densely hued wine has aromas of black plu...   
22            US  The aromas are brooding, with notes of barrel ...   
23        France  This fat, yeasty Champagne is comprised predom...   
24            US  There are intriguing touches to the nose of th...   
25            US  A blend of Merlot and Cabernet Franc, this win...   
26        France  This cooperative, based in Aÿ, has benefited f...   
27            US  Big oak defines this robustly dense and extrac...   
28         Italy  Aromas of black-skinned fruit, leather, underb...   
29            US  Juicy plum, raspberry and pencil lead lead the...   
...          ...                                                ...   
54146      Spain  Dry apple and lemon aromas are clean and one s...   
54147     France  A fine value for a northern Rhône of this qual...   
54148         US  A blend of Sauvignon Blanc, Sauvignon Gris and...   
54149         US  Hailing from an extreme coastal vineyard, this...   
54150         US  From the Ranch House block of the famous viney...   
54151         US  This is taut and dense, and requires time and ...   
54152         US  The third vintage of Mith is the best yet, com...   
54153  Argentina  This dark, meaty Malbec has some Agrelo fruit ...   
54154         US  This ripe, rich, almost decadently thick wine ...   
54155  Argentina  There is a select group of under-$20 Malbecs f...   
54156  Argentina  A lively, well-made blend of Tempranillo, Malb...   
54157         US  This pure Syrah from Reininger's estate vineya...   
54158      Spain  This represents Torres' highest effort in Prio...   
54159      Spain  Arguably on the heavy and rich side, but if yo...   
54160  Argentina  Andeluna's top wines tend to be ripe and plump...   
54161     France  A blend from several premier cru vineyards in ...   
54162  Argentina  Compared to the regular 2006 Malbec from Chaka...   
54163      Italy  A blend of Nero d'Avola and Syrah, this convey...   
54164         US  Hailing from one of the more popular vineyards...   
54165      Italy  A blend of 65% Cabernet Sauvignon, 30% Merlot ...   
54166  Argentina  Raspberry and cassis aromas are fresh and upri...   
54167         US  There's no bones about the use of oak in this ...   
54168         US  This opens with herbaceous dollops of thyme an...   
54169     France  Hugely spicy this rich wine is described as sw...   
54170      Spain  Lightly baked b

In [17]:
df_cleaned3.isna().sum()

country           0
description       0
designation       0
points            0
price             0
province          0
region_1          0
taster_name       0
title             0
variety           0
winery            0
color          7828
dtype: int64

In [18]:
df_cleaned4 = df_cleaned3.dropna()
df_cleaned4.color.describe()

count     46348
unique        3
top         Red
freq      30919
Name: color, dtype: object

In [19]:
df_cleaned4.groupby('color').count()


country  description  designation  points  price  province  region_1  \
color                                                                          
Orange      655          655          655     655    655       655       655   
Red       30919        30919        30919   30919  30919     30919     30919   
White     14774        14774        14774   14774  14774     14774     14774   

        taster_name  title  variety  winery  
color                                        
Orange          655    655      655     655  
Red           30919  30919    30919   30919  
White         14774  14774    14774   14774

In [20]:
#Dropping orange to do log reg
to_drop = ['Orange']
df_cleaned5 = df_cleaned4[~df_cleaned4['color'].isin(to_drop)]
df_cleaned5.groupby('color').count()

country  description  designation  points  price  province  region_1  \
color                                                                         
Red      30919        30919        30919   30919  30919     30919     30919   
White    14774        14774        14774   14774  14774     14774     14774   

       taster_name  title  variety  winery  
color                                       
Red          30919  30919    30919   30919  
White        14774  14774    14774   14774

In [21]:
df_cleaned5.describe(include= 'all')

country                                        description designation  \
count    45693                                              45693       45693   
unique       7                                              41833       20363   
top         US  Cigar box, café au lait, and dried tobacco aro...     Reserve   
freq     23255                                                  3        1095   
mean       NaN                                                NaN         NaN   
std        NaN                                                NaN         NaN   
min        NaN                                                NaN         NaN   
25%        NaN                                                NaN         NaN   
50%        NaN                                                NaN         NaN   
75%        NaN                                                NaN         NaN   
max        NaN                                                NaN         NaN   

              points         price    province              region_1  \
count   45693.000000  45693.000000       45693                 45693   
unique           NaN           NaN          62                   906   
top              NaN           NaN  California  Columbia Valley (WA)   
freq             NaN           NaN       12667                  2240   
mean       89.082266     41.385748         NaN                   NaN   
std         2.968545     37.391714         NaN                   NaN   
min        80.000000      4.000000         NaN                   NaN   
25%        87.000000     20.000000         NaN                   NaN   
50%        89.000000     32.000000         NaN                   NaN   
75%        91.000000     50.000000         NaN                   NaN   
max       100.000000   2013.000000         NaN                   NaN   

       taster_name                                        title     variety  \
count        45693                                        45693       45693   
unique          17                                        41334         263   
top     Roger Voss  Segura Viudas NV Extra Dry Sparkling (Cava)  Pinot Noir   
freq          7271                                            8        6736   
mean           NaN                                          NaN         NaN   
std            NaN                                          NaN         NaN   
min            NaN                                          NaN         NaN   
25%            NaN                                          NaN         NaN   
50%            NaN                                          NaN         NaN   
75%            NaN                                          NaN         NaN   
max            NaN                                          NaN         NaN   

                winery  color  
count            45693  45693  
unique            7725      2  
top     Columbia Crest    Red  
freq               146  30919  
mean               NaN    NaN  
std                NaN    NaN  
min                NaN    NaN  
25%                NaN    NaN  
50%                NaN    NaN  
75%                NaN    NaN  
max                NaN    NaN

# Potential target variables to Predict based on data above:
- color: there are 3 unique colors
- country: there are 7 unique countries
- points - below mean or above mean?
- price - below mean or above mean?
- taster name - based on the language they used?
- the variety of wine? - pinot noir etc (there are 434 total wine varieties)

## 1. Logistic Regression - Using StatsModel

In [24]:
df_cleaned6= pd.get_dummies(df_cleaned5, columns=['color'],drop_first = True)

In [27]:
from patsy import dmatrices

y, X = dmatrices('color_White ~ country  + points + price + province + taster_name',
                  df_cleaned6, return_type = "dataframe")

In [28]:
#fit model
import statsmodels.api as sm
logit_model = sm.Logit(y, X)
result = logit_model.fit()

         Current function value: 0.476539
         Iterations: 35


/anaconda3/lib/python3.6/site-packages/statsmodels/base/model.py:508: ConvergenceWarning: Maximum Likelihood optimization failed to converge. Check mle_retvals
  "Check mle_retvals", ConvergenceWarning)


In [29]:
result.summary()

/anaconda3/lib/python3.6/site-packages/statsmodels/base/model.py:1092: RuntimeWarning: invalid value encountered in sqrt
  bse_ = np.sqrt(np.diag(self.cov_params()))
/anaconda3/lib/python3.6/site-packages/scipy/stats/_distn_infrastructure.py:879: RuntimeWarning: invalid value encountered in greater
  return (self.a < x) & (x < self.b)
/anaconda3/lib/python3.6/site-packages/scipy/stats/_distn_infrastructure.py:879: RuntimeWarning: invalid value encountered in less
  return (self.a < x) & (x < self.b)
/anaconda3/lib/python3.6/site-packages/scipy/stats/_distn_infrastructure.py:1821: RuntimeWarning: invalid value encountered in less_equal
  cond2 = cond0 & (x <= self.a)


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:            color_White   No. Observations:                45693
Model:                          Logit   Df Residuals:                    45613
Method:                           MLE   Df Model:                           79
Date:                Sun, 06 Jan 2019   Pseudo R-squ.:                  0.2428
Time:                        12:28:45   Log-Likelihood:                -21774.
converged:                      False   LL-Null:                       -28757.
                                        LLR p-value:                     0.000
=====================================================================================================
                                        coef    std err          z      P>|z|      [0.025      0.975]
-----------------------------------------------------------------------------------------------------
Intercept                            -0.0459   9.42e+05  -4.88e-08      1.000   -1.85e+06    1.85e+06
country[T.Australia]                 -0.6545   3.52e+06  -1.86e-07      1.000   -6.89e+06    6.89e+06
country[T.Canada]                     0.2787        nan        nan        nan         nan         nan
country[T.France]                     0.9980    9.4e+05   1.06e-06      1.000   -1.84e+06    1.84e+06
country[T.Italy]                     -0.6054        nan        nan        nan         nan         nan
country[T.Spain]                     -0.2076        nan        nan        nan         nan         nan
country[T.US]                         2.7635    2.3e+06    1.2e-06      1.000    -4.5e+06     4.5e+06
province[T.Andalucia]                 0.0843        nan        nan        nan         nan         nan
province[T.Arizona]                  -2.2793   2.69e+06  -8.48e-07      1.000   -5.27e+06    5.27e+06
province[T.Australia Other]          -0.0914      4e+06  -2.28e-08      1.000   -7.85e+06    7.85e+06
province[T.Beaujolais]               20.1230   1.85e+04      0.001      0.999   -3.63e+04    3.63e+04
province[T.Bordeaux]                 -3.4009      0.191    -17.813      0.000      -3.775      -3.027
province[T.British Columbia]         -1.2479        nan        nan        nan         nan         nan
province[T.Burgundy]                 -0.6319      0.181     -3.486      0.000      -0.987      -0.277
province[T.California]               -5.2839   2.69e+06  -1.97e-06      1.000   -5.27e+06    5.27e+06
province[T.Catalonia]                 1.1899        nan        nan        nan         nan         nan
province[T.Central Italy]             0.8169        nan        nan        nan         nan         nan
province[T.Central Spain]            -1.1118        nan        nan        nan         nan         nan
province[T.Champagne]                 2.9057      0.239     12.172      0.000       2.438       3.374
province[T.Colorado]                 -3.7036   2.69e+06  -1.38e-06      1.000   -5.27e+06    5.27e+06
province[T.Connecticut]              -2.9972   2.69e+06  -1.12e-06      1.000   -5.27e+06    5.27e+06
province[T.France Other]             -1.6320      0.216     -7.551      0.000      -2.056      -1.208
province[T.Galicia]                   2.2062        nan        nan        nan         nan         nan
province[T.Hawaii]                   19.1112   2.69e+06   7.11e-06      1.000   -5.27e+06    5.27e+06
province[T.Idaho]                    -3.6721   2.69e+06  -1.37e-06      1.000   -5.27e+06    5.27e+06
province[T.Illinois]                 16.2204   2.69e+06   6.04e-06      1.000   -5.27e+06    5.27e+06
province[T.Iowa]                     24.3074   2.83e+06    8.6e-06      1.000   -5.54e+06    5.54e+06
province[T.Italy Other]               2.5391        nan        nan        nan         nan         nan
province[T.Kentucky]                -14.0247   2.69e+06  -5.22e-06      1.000  

In [30]:
y, X = dmatrices('color_White ~ country  + points + price + taster_name',
                  df_cleaned6, return_type = "dataframe")
logit_model = sm.Logit(y, X)
result = logit_model.fit()
result.summary()

         Current function value: 0.548269
         Iterations: 35


/anaconda3/lib/python3.6/site-packages/statsmodels/base/model.py:508: ConvergenceWarning: Maximum Likelihood optimization failed to converge. Check mle_retvals
  "Check mle_retvals", ConvergenceWarning)


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:            color_White   No. Observations:                45693
Model:                          Logit   Df Residuals:                    45668
Method:                           MLE   Df Model:                           24
Date:                Sun, 06 Jan 2019   Pseudo R-squ.:                  0.1288
Time:                        12:28:46   Log-Likelihood:                -25052.
converged:                      False   LL-Null:                       -28757.
                                        LLR p-value:                     0.000
=====================================================================================================
                                        coef    std err          z      P>|z|      [0.025      0.975]
-----------------------------------------------------------------------------------------------------
Intercept                             1.9475      0.537      3.629      0.000       0.896       2.999
country[T.Australia]                 -1.0408      0.342     -3.040      0.002      -1.712      -0.370
country[T.Canada]                     0.5174      0.383      1.352      0.176      -0.232       1.267
country[T.France]                    -0.5923      0.339     -1.749      0.080      -1.256       0.071
country[T.Italy]                      0.5839      0.302      1.932      0.053      -0.008       1.176
country[T.Spain]                      0.6920      0.067     10.276      0.000       0.560       0.824
country[T.US]                        -0.9021      0.346     -2.606      0.009      -1.581      -0.224
taster_name[T.Anna Lee C. Iijima]     0.7012      0.145      4.841      0.000       0.417       0.985
taster_name[T.Anne Krebiehl MW]       1.8736      0.271      6.905      0.000       1.342       2.405
taster_name[T.Carrie Dykes]          -0.5106      0.257     -1.986      0.047      -1.015      -0.007
taster_name[T.Christina Pickard]      0.2102      1.432      0.147      0.883      -2.597       3.017
taster_name[T.Fiona Adams]          -19.3050   6124.845     -0.003      0.997    -1.2e+04     1.2e+04
taster_name[T.Jim Gordon]            -1.0530      0.142     -7.414      0.000      -1.331      -0.775
taster_name[T.Joe Czerwinski]        -0.4877      0.231     -2.110      0.035      -0.941      -0.035
taster_name[T.Kerin O’Keefe]         -2.6466      0.327     -8.086      0.000      -3.288      -2.005
taster_name[T.Lauren Buzzeo]         -0.2231      0.277     -0.805      0.421      -0.766       0.320
taster_name[T.Matt Kettmann]         -0.8736      0.139     -6.272      0.000      -1.147      -0.601
taster_name[T.Michael Schachner]     -2.7304      0.366     -7.450      0.000      -3.449      -2.012
taster_name[T.Paul Gregutt]          -1.2253      0.138     -8.898      0.000      -1.495      -0.955
taster_name[T.Roger Voss]             0.3344      0.243      1.374      0.170      -0.143       0.812
taster_name[T.Sean P. Sullivan]      -1.2822      0.141     -9.083      0.000      -1.559      -1.006
taster_name[T.Susan Kostrzewa]        0.1264      0.217      0.582      0.560      -0.299       0.552
taster_name[T.Virginie Boone]        -0.8226      0.137     -6.010      0.000      -1.091      -0.554
points                               -0.0072      0.005     -1.577      0.115      -0.016       0.002
price                                -0.0132      0.001    -25.191      0.000      -0.014      -0.012
=====================================================================================================
"""

In [31]:
y, X = dmatrices('color_White ~ country + price + taster_name',
                  df_cleaned6, return_type = "dataframe")
logit_model = sm.Logit(y, X)
result = logit_model.fit()
result.summary()

         Current function value: 0.548296
         Iterations: 35


/anaconda3/lib/python3.6/site-packages/statsmodels/base/model.py:508: ConvergenceWarning: Maximum Likelihood optimization failed to converge. Check mle_retvals
  "Check mle_retvals", ConvergenceWarning)


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:            color_White   No. Observations:                45693
Model:                          Logit   Df Residuals:                    45669
Method:                           MLE   Df Model:                           23
Date:                Sun, 06 Jan 2019   Pseudo R-squ.:                  0.1288
Time:                        12:28:47   Log-Likelihood:                -25053.
converged:                      False   LL-Null:                       -28757.
                                        LLR p-value:                     0.000
=====================================================================================================
                                        coef    std err          z      P>|z|      [0.025      0.975]
-----------------------------------------------------------------------------------------------------
Intercept                             1.3360      0.371      3.600      0.000       0.609       2.063
country[T.Australia]                 -1.0521      0.342     -3.072      0.002      -1.723      -0.381
country[T.Canada]                     0.5171      0.383      1.351      0.177      -0.233       1.267
country[T.France]                    -0.5985      0.339     -1.765      0.078      -1.263       0.066
country[T.Italy]                      0.5785      0.302      1.915      0.056      -0.014       1.171
country[T.Spain]                      0.6898      0.067     10.248      0.000       0.558       0.822
country[T.US]                        -0.8994      0.346     -2.597      0.009      -1.578      -0.221
taster_name[T.Anna Lee C. Iijima]     0.6856      0.145      4.744      0.000       0.402       0.969
taster_name[T.Anne Krebiehl MW]       1.8461      0.271      6.814      0.000       1.315       2.377
taster_name[T.Carrie Dykes]          -0.5170      0.257     -2.011      0.044      -1.021      -0.013
taster_name[T.Christina Pickard]      0.1833      1.432      0.128      0.898      -2.623       2.990
taster_name[T.Fiona Adams]          -19.7745   7737.161     -0.003      0.998   -1.52e+04    1.51e+04
taster_name[T.Jim Gordon]            -1.0769      0.141     -7.624      0.000      -1.354      -0.800
taster_name[T.Joe Czerwinski]        -0.4945      0.231     -2.139      0.032      -0.948      -0.041
taster_name[T.Kerin O’Keefe]         -2.6573      0.327     -8.120      0.000      -3.299      -2.016
taster_name[T.Lauren Buzzeo]         -0.2254      0.277     -0.813      0.416      -0.769       0.318
taster_name[T.Matt Kettmann]         -0.9030      0.138     -6.542      0.000      -1.174      -0.632
taster_name[T.Michael Schachner]     -2.7385      0.367     -7.469      0.000      -3.457      -2.020
taster_name[T.Paul Gregutt]          -1.2481      0.137     -9.112      0.000      -1.517      -0.980
taster_name[T.Roger Voss]             0.3236      0.243      1.329      0.184      -0.154       0.801
taster_name[T.Sean P. Sullivan]      -1.3032      0.141     -9.271      0.000      -1.579      -1.028
taster_name[T.Susan Kostrzewa]        0.1332      0.217      0.614      0.539      -0.292       0.558
taster_name[T.Virginie Boone]        -0.8422      0.136     -6.178      0.000      -1.109      -0.575
price                                -0.0136      0.000    -29.721      0.000      -0.014      -0.013
=====================================================================================================
"""

## 2. Logistic Regression - Using SciKit Learn

In [32]:
X = df_cleaned6[df_cleaned6.columns[:-1]]
y = df_cleaned6.color_White
X.head()

country                                        description  \
0      US  Pineapple rind, lemon pith and orange blossom ...   
1      US  Much like the regular bottling from 2012, this...   
2   Spain  Blackberry and raspberry aromas show a typical...   
3   Italy  Here's a bright, informal red that opens with ...   
5      US  Soft, supple plum envelopes an oaky structure ...   

                          designation  points  price           province  \
0                Reserve Late Harvest      87   13.0           Michigan   
1  Vintner's Reserve Wild Child Block      87   65.0             Oregon   
2                        Ars In Vitro      87   15.0     Northern Spain   
3                             Belsito      87   16.0  Sicily & Sardinia   
5                      Mountain Cuvée      87   19.0         California   

              region_1         taster_name  \
0  Lake Michigan Shore  Alexander Peartree   
1    Willamette Valley        Paul Gregutt   
2              Navarra   Michael Schachner   
3             Vittoria       Kerin O’Keefe   
5          Napa Valley      Virginie Boone   

                                               title             variety  \
0  St. Julian 2013 Reserve Late Harvest Riesling ...            Riesling   
1  Sweet Cheeks 2012 Vintner's Reserve Wild Child...          Pinot Noir   
2  Tandem 2011 Ars In Vitro Tempranillo-Merlot (N...  Tempranillo-Merlot   
3   Terre di Giurfo 2013 Belsito Frappato (Vittoria)            Frappato   
5  Kirkland Signature 2011 Mountain Cuvée Caberne...  Cabernet Sauvignon   

               winery  
0          St. Julian  
1        Sweet Cheeks  
2              Tandem  
3     Terre di Giurfo  
5  Kirkland Signature

In [37]:
price = X['price']
for row in price:
    normalized_price = (price-price.min())/ (price.max()-price.min())

normalized_price.head()

0    0.004480
1    0.030363
2    0.005475
3    0.005973
5    0.007466
Name: price, dtype: float64

In [41]:
for col in (df_cleaned6[['price','points']]):
    df_cleaned6[col] = (df_cleaned6[col]-min(df_cleaned6[col]))/ (max(df_cleaned6[col]) - min(df_cleaned6[col]))
df_cleaned6.head()

country                                        description  \
0      US  Pineapple rind, lemon pith and orange blossom ...   
1      US  Much like the regular bottling from 2012, this...   
2   Spain  Blackberry and raspberry aromas show a typical...   
3   Italy  Here's a bright, informal red that opens with ...   
5      US  Soft, supple plum envelopes an oaky structure ...   

                          designation  points     price           province  \
0                Reserve Late Harvest    0.35  0.004480           Michigan   
1  Vintner's Reserve Wild Child Block    0.35  0.030363             Oregon   
2                        Ars In Vitro    0.35  0.005475     Northern Spain   
3                             Belsito    0.35  0.005973  Sicily & Sardinia   
5                      Mountain Cuvée    0.35  0.007466         California   

              region_1         taster_name  \
0  Lake Michigan Shore  Alexander Peartree   
1    Willamette Valley        Paul Gregutt   
2              Navarra   Michael Schachner   
3             Vittoria       Kerin O’Keefe   
5          Napa Valley      Virginie Boone   

                                               title             variety  \
0  St. Julian 2013 Reserve Late Harvest Riesling ...            Riesling   
1  Sweet Cheeks 2012 Vintner's Reserve Wild Child...          Pinot Noir   
2  Tandem 2011 Ars In Vitro Tempranillo-Merlot (N...  Tempranillo-Merlot   
3   Terre di Giurfo 2013 Belsito Frappato (Vittoria)            Frappato   
5  Kirkland Signature 2011 Mountain Cuvée Caberne...  Cabernet Sauvignon   

               winery  color_White  
0          St. Julian            1  
1        Sweet Cheeks            0  
2              Tandem            0  
3     Terre di Giurfo            0  
5  Kirkland Signature            0

In [51]:
X = df_cleaned6[['price','points']]
y = df_cleaned6.color_White
X.head()

price  points
0  0.004480    0.35
1  0.030363    0.35
2  0.005475    0.35
3  0.005973    0.35
5  0.007466    0.35

In [52]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split

In [53]:
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=0)
logreg = LogisticRegression(fit_intercept = False, C = 1e12) #Starter code


In [54]:
model_log = logreg.fit(X_train, y_train)
model_log

/anaconda3/lib/python3.6/site-packages/sklearn/linear_model/logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


LogisticRegression(C=1000000000000.0, class_weight=None, dual=False,
          fit_intercept=False, intercept_scaling=1, max_iter=100,
          multi_class='warn', n_jobs=None, penalty='l2', random_state=None,
          solver='warn', tol=0.0001, verbose=0, warm_start=False)

In [55]:
y_hat_test = logreg.predict(X_test)
y_hat_train = logreg.predict(X_train)

In [57]:
residuals = y_train - y_hat_train
print(pd.Series(residuals).value_counts())
print(pd.Series(residuals).value_counts(normalize=True))

0    23111
1    11158
Name: color_White, dtype: int64
0    0.6744
1    0.3256
Name: color_White, dtype: float64


### How many times was the classifier correct for the training set?  
23111 correct, 67.44% accuracy

In [58]:
residuals = y_test - y_hat_test
print(pd.Series(residuals).value_counts())
print(pd.Series(residuals).value_counts(normalize=True))
#63 correct, 83% accuracy

0    7808
1    3616
Name: color_White, dtype: int64
0    0.683473
1    0.316527
Name: color_White, dtype: float64


### How many times was the classifier correct for the test set?
7808 correct, 68.3473% accuracy